Deprecate unnecessary escape characters

ruudk opened the pull request:
This is a first attempt at solving #4123 and #2712.

Currently it writes the deprecations to an array in the Lexer. This is probably not the way to do it. Should we directly trigger `E_USER_DEPRECATED` errors?

/cc @stof @fabpot Let me know what you think 😊

ruudk commented on src/Lexer.php:
```diff
@@ -396,6 +402,64 @@ private function lexExpression(): void
+                    }
+                    $result .= $nextChar;
+                    $i++;
+                } elseif ($nextChar === 'x' && $i + 2 < $length && ctype_xdigit($str[$i + 1]) && ctype_xdigit($str[$i + 2])) {
+                    $result .= chr(hexdec($str[$i + 1] . $str[$i + 2]));
+                    $i += 2;
```
Not sure if this is correct. Does Twig even support it? https://github.com/twigphp/Twig/issues/3604

stof commented:
To me, this should directly trigger `E_USER_DEPRECATED` (or use the `trigger_deprecation` of the symfony/deprecation-contracts if we already use it as the way to trigger deprecations).

And this should go in the 3.x branch (so that 4.0 can ship with the new behavior instead of having to wait until 5.0)

stof commented on tests/LexerTest.php:
```diff
@@ -178,23 +178,97 @@ public function testBigNumbers()
+            App\Test
+            EOF,
+            [
+                'Character \ at position 5 does not require to be escaped.'
```
This deprecation looks weird to me. Escaping a `\` should become optional, but should stay allowed IMO (like in PHP).

ruudk replied:
Removed it.

Can we somehow notify the developer in v4 that escaping a `\` is not needed? Maybe via `E_USER_NOTICE`? Or should we just not care about that 😊

stof replied:
we should not care (just like PHP allows escaping them). Especially given that escaping them is still required in some cases

ruudk replied:
Ok clear!

stof commented on tests/LexerTest.php:
```diff
@@ -178,23 +178,97 @@ public function testBigNumbers()
+            {{ 'App\Test' }}
+            TWIG,
+            <<<'EOF'
+            App\Test
```
This cannot be the behavior in a version *deprecating* the unnecessary escaping. The result of parsing must stay the same when introducing the deprecation (we should warn about a *future* behavior change, otherwise we are already breaking BC and the deprecation is useless).

Here is a good rule of thumb: except for the assertion on the deprecation message, there should be no difference in the tests between making them pass on the current stable 3.x release and the version containing the deprecation.

ruudk replied:
Changed it back to the old behavior. The test now reflects what happens when you don't escape `\`.

ruudk commented:
Thanks for the review. I think it makes sense. I will change this PR to target v3 and make sure we only deprecate and not introduce new functionality.

Then for v4 we can do the other change.

ruudk committed:
Deprecate unnecessary escape characters

See #4123 #2712

This allows to change the implementation in v4 so that we no longer have to escape backslashes.

ruudk commented:
Oke, I updated the PR to target V3. We don't change the behavior. We only deprecate, using `trigger_deprecation`, when we detect an unnecessary escape character. 

When this is merged, v4 is rebased with v3, I can provide the PR that changes the behavior there. 

Please have another look when you have time 😊

fabpot commented on src/Lexer.php:
```diff
@@ -382,6 +382,62 @@ private function lexExpression(): void
+                    $i++;
+                } elseif ($nextChar === "'" || $nextChar === '"') {
+                    if ($nextChar !== $quoteType) {
+                        trigger_deprecation('twig/twig', '3.10', "Character %s at position %d does not require to be escaped.", $nextChar, $i + 2);
```
```suggestion
                        trigger_deprecation('twig/twig', '3.11', "Character "%s" at position %d does not need to be escaped anymore.", $nextChar, $i + 2);
```

fabpot commented on src/Lexer.php:
```diff
@@ -382,6 +382,62 @@ private function lexExpression(): void
+                    $result .= chr(octdec($octal));
+                    $i++;
+                } else {
+                    trigger_deprecation('twig/twig', '3.10', "Character %s at position %d does not require to be escaped.", $nextChar, $i + 2);
```
```suggestion
                    trigger_deprecation('twig/twig', '3.11', "Character "%s" at position %d does not need to be escaped anymore.", $nextChar, $i + 2);
```

fabpot reviewed (changes requested):
We should probably add a note in the docs explaining when one needs to escape.

ruudk commented:
I refactored the code a bit and added more test cases. I think good to carefully review these test cases. Let me know if you think I can add more cases. 

Also updated the docs.

ruudk commented on doc/templates.rst:
```diff
@@ -526,7 +526,18 @@ exist:
+   * ``\r``: Carriage return
+   * ``\t``: Horizontal tab
+   * ``\v``: Vertical tab
+   * ``\#``: Hash
```
Not sure about this one. Maybe this should only apply when `#{` is used?

ruudk commented on src/Lexer.php:
```diff
@@ -382,6 +382,58 @@ private function lexExpression(): void
+                    }
+                    $result .= $nextChar;
+                    $i++;
+                } elseif ($nextChar === 'x' && $i + 2 < $length && ctype_xdigit($str[$i + 1]) && ctype_xdigit($str[$i + 2])) {
+                    $result .= chr(hexdec($str[$i + 1] . $str[$i + 2]));
+                    $i += 2;
```
Not sure about this. If we need to handle this, we should also provide test cases for it.

stof replied:
we indeed should not add support for things without testing them. But even more important, we should not change the behavior of strings in 3.x and so any test you add must first be run on the existing 3.x version (ensuring we keep the same behavior)

ruudk replied:
Added tests for it, and made sure it stays the same. 

Also found some PHP tests for stripcslashes https://github.com/php/php-src/blob/master/ext/standard/tests/strings/add-and-stripcslashes.phpt

ruudk commented on tests/LexerTest.php:
```diff
@@ -178,23 +181,70 @@ public function testBigNumbers()
+            '{{ \'App\\\\Test\' }}',
+            'App\\Test',
+        ];
+        yield '{{ \'App\Test\' }} => AppTest' => [
+            '{{ \'App\\Test\' }}',
+            'AppTest',
+        ];
```
This should trigger a deprecation, as it doesn't do anything. In v4 this will become App\Test

ruudk commented on tests/LexerTest.php:
```diff
@@ -178,23 +181,70 @@ public function testBigNumbers()
+            '{{ \'App\\Test\' }}',
+            'AppTest',
+        ];
+        yield '{{ \'A\B\C\' }} => ABC' => [
+            '{{ \'A\B\C\' }}',
+            'ABC',
+        ];
```
This should trigger a deprecation, as it doesn't do anything. In v4 this will become A\B\C

ruudk commented on src/Lexer.php:
```diff
@@ -382,6 +382,62 @@ private function lexExpression(): void
+                    }
+                    $result .= $nextChar;
+                    $i++;
+                } elseif ($nextChar === '#' && $i + 2 < $length && $str[$i + 2] === '{') {
+                    $result .= '#{';
+                    $i += 2;
```
I added this case, to make sure that escaping string Interpolation works. 

Should we also document this?

fabpot replied:
As escaping string interpolation is already supported, this code is needed to avoid the deprecation in case a string interpolation is escaped, right?

Escaping a string interpolation is tested here: `LexerTest::testStringWithEscapedInterpolation()` and I've just submitted #4189 for the missing documentation.

ruudk replied:
> this code is needed to avoid the deprecation in case a string interpolation is escaped, right

Exactly.

ruudk commented:
Alright, I think I'm done. I added more documentation related to Octal and Hexadecimal escape sequences.

fabpot commented on doc/templates.rst:
```diff
@@ -527,6 +527,18 @@ exist:
+   * ``\``: Backslash
+  
+   When using single-quoted strings, the single quote character (``'``) needs to be escaped with a backslash (``\'``).
+   When using double-quoted strings, the double quote character (``"``) needs to be escaped with a backslash (``\"``).
```
Can you add an empty line?

ruudk replied:
Done 😊

ruudk commented:
@fabpot what do you think of this? Anything we should change or add more tests?

fabpot commented on doc/templates.rst:
```diff
@@ -527,6 +527,19 @@ exist:
+   * ``\0`` to ``\377``: Octal escape sequences representing characters
+   * ``\``: Backslash
+  
+   When using single-quoted strings, the single quote character (``'``) needs to be escaped with a backslash (``\'``).
```
This is already explained in the first paragraph, maybe we should remove the explanation from there and add an example here like done there :)

ruudk replied:
Done, is this better?

fabpot commented on src/Lexer.php:
```diff
@@ -382,6 +382,74 @@ private function lexExpression(): void
+                    $i++;
+                }
+            } else {
+                $result .= $str[$i];
```
Can we negate the condition so that this becomes the default and use `continue` there so that we can unindent the big block?

ruudk replied:
Done!

fabpot commented:
I've just run a small performance test via the script below and (maybe as expected) the hit can be significant:

```php
<?php

require_once __DIR__.'/vendor/autoload.php';

$string = str_repeat("日本では、春になると桜の花が咲きます。多くの人々は、公園や川の近くに集まり、お花見を楽しみます。桜の花びらが風に舞い、まるで雪のように見える瞬間は、とても美しいです。", 10);
$loader = new Twig\Loader\ArrayLoader([
    'index.twig' => <<<EOF
{{ "$string" }}
EOF
    ,
]);

$twig = new Twig\Environment($loader);

for ($i = 0; $i < 10000; $i++) {
    $twig->tokenize(new Twig\Source($twig->getLoader()->getSourceContext('index.twig')->getCode(), 'index.twig'));
}
```

On my machine, ~100ms for the 3.x branch vs ~900ms for this branch.

But I suppose that this kind of long string is not very common. I would expect "hardcoded" strings to be quite short in templates.

ruudk commented:
Maybe we could optimize the code a bit. First check if the string contains a backslash, then run the method. Otherwise don't do anything?

fabpot commented:
> Maybe we could optimize the code a bit. First check if the string contains a backslash, then run the method. Otherwise don't do anything?

Sure, as we have a perf script now, we can try a few obvious strategies to see their impact. But then again, let's optimize for short strings first and make sure that long strings are not too slow to lex.

ruudk commented:
I ran the bench on my machine:

```
// v3 branch
Benchmark 1: php bench.php
  Time (mean ± σ):     110.2 ms ±   3.7 ms    [User: 98.5 ms, System: 7.4 ms]
  Range (min … max):   108.6 ms … 126.9 ms    23 runs

// this branch
Benchmark 1: php bench.php
  Time (mean ± σ):      1.018 s ±  0.010 s    [User: 0.991 s, System: 0.009 s]
  Range (min … max):    1.007 s …  1.032 s    10 runs
```

When I change the `stripcslashes` method:
```diff
 private function stripcslashes(string $str, string $quoteType): string
 {
+    if (!str_contains($str, '\\')) {
+        return $str;
+    }
+    // ...
```

it becomes this:
```
// this branch
Benchmark 1: php bench.php
  Time (mean ± σ):      95.7 ms ±   1.1 ms    [User: 84.9 ms, System: 7.5 ms]
  Range (min … max):    94.5 ms …  99.0 ms    29 runs
```

But obviously, that's because the bench does not contain any escape characters. 

So let's change the bench to:
```php
<?php

require_once __DIR__.'/vendor/autoload.php';

$string = str_repeat("日本では、春になると桜の花が咲きます。\n多くの人々は、公園や川の近くに集まり、お花見を楽しみます。\\n桜の花びらが風に舞い、まるで雪のように見える瞬間は、とても美しいです。", 10);
$loader = new Twig\Loader\ArrayLoader([
    'index.twig' => <<<EOF
{{ "$string" }}
EOF
    ,
]);

$twig = new Twig\Environment($loader);

for ($i = 0; $i < 10000; $i++) {
    $twig->tokenize(new Twig\Source($twig->getLoader()->getSourceContext('index.twig')->getCode(), 'index.twig'));
}
```

Now we get
```
// this branch
Benchmark 1: php bench.php
  Time (mean ± σ):      1.026 s ±  0.002 s    [User: 0.999 s, System: 0.008 s]
  Range (min … max):    1.024 s …  1.029 s    10 runs
```

So the `str_contains` check is making it a bit slower when there is a backslash found. But when there is not, it can return the raw string directly, making it even faster than v3.

fabpot commented on src/Lexer.php:
```diff
@@ -382,6 +382,75 @@ private function lexExpression(): void
+                $result .= chr(octdec($octal));
+                $i++;
+            } else {
+                trigger_deprecation('twig/twig', '3.11', 'Character "%s" at position %d does not need to be escaped anymore.', $nextChar, $i + 2);
```
```suggestion
                trigger_deprecation('twig/twig', '3.12', 'Character "%s" at position %d does not need to be escaped anymore.', $nextChar, $i + 2);
```

fabpot commented on src/Lexer.php:
```diff
@@ -382,6 +382,75 @@ private function lexExpression(): void
+                $i++;
+            }
+        }
+        return $result;
```
missing empty line above

fabpot commented on src/Lexer.php:
```diff
@@ -382,6 +382,75 @@ private function lexExpression(): void
+        for ($i = 0; $i < $length; $i++) {
+            if (!($str[$i] === '\\' && $i + 1 < $length)) {
+                $result .= $str[$i];
+                continue;
```
missing empty line above

fabpot commented on src/Lexer.php:
```diff
@@ -382,6 +382,75 @@ private function lexExpression(): void
+        ];
+
+        for ($i = 0; $i < $length; $i++) {
+            if (!($str[$i] === '\\' && $i + 1 < $length)) {
```
```suggestion
            if ($str[$i] !== '\\' || $i + 1 >= $length) {
```

fabpot reviewed (changes requested):
You haven't added the `str_contains()` call yet, right?

ruudk commented:
@fabpot Applied the feedback and also added the `str_contains` optimization.

fabpot commented:
Thank you @ruudk.

ruudk commented:
Thanks! I will prep the next PR when v4 is synced.

smnandre commented:
Hi @ruudk ... i may have a version that reduces the runtime difference, using chunks to avoid looping on every char.

It passes the tests, but i'm not sure if this goes in the direction you wanted for it.

So if you want to check / benchmark it: 

```php
private function stripcslashes(string $str, string $quoteType): string
{
    if (!str_contains($str, '\\')) {
        return $str;
    }

    $result = '';
    $length = strlen($str);
    $specialChars = [
        'f' => "\f",
        'n' => "\n",
        'r' => "\r",
        't' => "\t",
        'v' => "\v",
    ];

    $i = 0;
    while ($i < $length) {
        if (false === $pos = strpos($str, '\\', $i)) {
            $result .= substr($str, $i);
            break;
        }

        $result .= substr($str, $i, $pos - $i);
        $i = $pos + 1;

        if ($i >= $length) {
            $result .= '\\';
            break;
        }

        $nextChar = $str[$i];

        if (isset($specialChars[$nextChar])) {
            $result .= $specialChars[$nextChar];
        } elseif ($nextChar === '\\' || $nextChar === $quoteType) {
            $result .= $nextChar;
        } elseif ($nextChar === '#' && $i + 1 < $length && $str[$i + 1] === '{') {
            $result .= '#{';
            $i++;
        } elseif ($nextChar === 'x' && $i + 1 < $length && ctype_xdigit($str[$i + 1])) {
            $hex = $str[++$i];
            if ($i + 1 < $length && ctype_xdigit($str[$i + 1])) {
                $hex .= $str[++$i];
            }
            $result .= chr(hexdec($hex));
        } elseif (ctype_digit($nextChar) && $nextChar < '8') {
            $octal = $nextChar;
            while ($i + 1 < $length && ctype_digit($str[$i + 1]) && $str[$i + 1] < '8' && strlen($octal) < 3) {
                $octal .= $str[++$i];
            }
            $result .= chr(octdec($octal));
        } else {
            trigger_deprecation('twig/twig', '3.12', sprintf('Character "%s" at position %d does not need to be escaped anymore.', $nextChar, $i + 1));
            $result .= $nextChar;
        }

        $i++;
    }

    return $result;
}
```

ruudk commented:
@smnandre smart! Will check it out today

ruudk commented:
Created the PR to optimize this further: https://github.com/twigphp/Twig/pull/4196
